akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala

```diff
@@ -0,0 +1,114 @@
+ * patterned after org.uncommons.maths.random by Daniel Dyer (Apache License 2.0)
+ *
+ * Non-linear random number generator based on the AES block cipher in counter mode.
+ * Uses the seed as a key to encrypt a 128-bit counter using AES(Rijndael).
+ *
+ * Keys larger than 128-bit for the AES cipher require
+ * the inconvenience of installing the unlimited strength cryptography policy
+ * files for the Java platform.  Larger keys may be used (192 or 256 bits) but if the
+ * cryptography policy files are not installed, a
+ * java.security.GeneralSecurityException will be thrown.
+ *
+ * NOTE: this class is not serializable
+ */
+
+private[akka] class AESCounterBuiltinRNG(val seed: Array[Byte]) extends Random {
+  import CounterRNGConstants._
+  private final val entropySource = new SecureRandom
+
+  // mutable state below, concurrent accesses need lock
+  private val lock = new ReentrantLock
+  private val counter: Array[Byte] = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  private var index: Int = 0
+  private var currentBlock: Array[Byte] = null
+  private var reseedFuture: Future[Array[Byte]] = null
+  private var bitsSinceSeeding: Long = 0
+
+  private val cipher = Cipher.getInstance("AES/CTR/NoPadding")
+  private val ivArr = Array.fill[Byte](COUNTER_SIZE_BYTES)(0)
+  ivArr(0) = (ivArr(0) + 1.toByte).toByte
+  private val ivSpec = new IvParameterSpec(ivArr)
+  cipher.init(Cipher.ENCRYPT_MODE, new this.AESKey(seed), ivSpec)
+
+  @Override
+  override protected def next(bits: Int): Int = {
+    try {
+      lock.lock()
+      bitsSinceSeeding += bits
+      if (currentBlock == null || currentBlock.length - index < 4) {
+        try {
+          currentBlock = cipher.doFinal(counter)
+          index = 0
+        } catch {
+          case ex: Exception ⇒
+            // Generally Cipher.doFinal() from nextBlock may throw various exceptions.
+            // However this should never happen.  If initialisation succeeds without exceptions
+            // we should be able to proceed indefinitely without exceptions.
+            throw new IllegalStateException("Failed creating next random block.", ex)
+        }
+      }
+      val result = (BITWISE_BYTE_TO_INT & currentBlock(index + 3)) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index + 2)) << 8) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index + 1)) << 16) |
+        ((BITWISE_BYTE_TO_INT & currentBlock(index)) << 24)
+
+      index += 4
+      result >>> (32 - bits)
+    } finally {
+      import scala.concurrent.ExecutionContext.Implicits.global
+      if (bitsSinceSeeding > RESEEDING_THRESHOLD) {
+        if (reseedFuture == null) {
+          //          println(s"reseeding started! bits: $bitsSinceSeeding")
+          reseedFuture = Future { entropySource.generateSeed(32) }
+        }
+        if (bitsSinceSeeding > RESEEDING_DEADLINE) {
+          //          println(s"blocking for the seeding result bits: $bitsSinceSeeding")
+          Await.ready(reseedFuture, Duration.Inf)
```

Comment:
I'm always scared about blocking forever on something because there is no sign of why the application is halting. Yeah, it should always return within reasonable time, but I'd rather have a long timeout (e.g. 5 minutes) with logging in case of failure than trusting an external "service".

Retrying might be overkill.